// @storybook/vue3/dist/cjs/server/framework-preset-vue3.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.previewAnnotations = void 0;
exports.webpack = webpack;

var _vueLoader = require("vue-loader");

var _webpack = require("webpack");

var _coreCommon = require("@storybook/core-common");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function webpack(config) {
  return _objectSpread(_objectSpread({}, config), {}, {
    plugins: [...config.plugins, new _vueLoader.VueLoaderPlugin(), new _webpack.DefinePlugin({
      __VUE_OPTIONS_API__: JSON.stringify(true),
      __VUE_PROD_DEVTOOLS__: JSON.stringify(true)
    })],
    module: _objectSpread(_objectSpread({}, config.module), {}, {
      rules: [...config.module.rules, {
        test: /\.vue$/,
        loader: require.resolve('vue-loader'),
        options: {}
      }, {
        test: /\.ts$/,
        use: [{
          loader: require.resolve('ts-loader'),
          options: {
            transpileOnly: true,
            appendTsSuffixTo: [/\.vue$/]
          }
        }]
      }, {
        test: /\.tsx$/,
        use: [{
          loader: require.resolve('ts-loader'),
          options: {
            transpileOnly: true,
            // Note this is different from the `appendTsSuffixTo` above!
            appendTsxSuffixTo: [/\.vue$/]
          }
        }]
      }]
    }),
    resolve: _objectSpread(_objectSpread({}, config.resolve), {}, {
      extensions: [...config.resolve.extensions, '.vue'],
      alias: _objectSpread(_objectSpread({}, config.resolve.alias), {}, {
        vue$: require.resolve('vue/dist/vue.esm-bundler.js')
      })
    })
  });
}

var previewAnnotations = function (entry = []) {
  return [...entry, (0, _coreCommon.findDistEsm)(__dirname, 'client/preview/config')];
};

exports.previewAnnotations = previewAnnotations;
